### PR TITLE
Update selected reserved roles for remote indices privileges

### PR DESCRIPTION
**Pull request by ywangd**
The new RCS model requires remote indices privileges to be specified separately from local indices privileges. This PR updates selected reserved roles with necessary remote indices privileges to ensure they work with both legacy and new RCS remote clusters.

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 1]
```diff
@@ -81,10 +82,36 @@ public class ReservedRolesStore implements BiConsumer<Set<String>, ActionListene
...
+                    "*"
+                ),
+                new RoleDescriptor.RemoteIndicesPrivileges(
+                    RoleDescriptor.IndicesPrivileges.builder()
+                        .indices("*")
+                        .privileges("monitor", "read", "view_index_metadata", "read_cross_cluster")
+                        .allowRestrictedIndices(true)
+                        .build(),
+                    "*"
+                ) }
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:99]: We can discuss whether `superuser` should only have `read` and `read_cross_cluster` privileges. They are the only supported ones by the new RCS. But the remote indices privileges itself does not prevent the usage of any index privileges, e.g. it is totally ok to define a custom role that has `all` privileges for any remote indices. 

So my view is that for the privileges itself, it do not need to care whether certain privilege is supported by new RCS, which is a downstream implementation detail. For the superuser role, it is much easier to say that the role has exactly the same indices privileges on any clusters and this suits what a `superuser` would imply - a builtin powerful role that is identical everywhere. 

Bottom line is that we don't have any security concern since the ultimate permissions are subject to the cluster API key.

**Comment by ywangd** [comment]
When everything is wired up, we can have REST tests that excercise CCS with reserved roles/users and service account. I plan to raise an issue for improving test coverage as there are a few other things that could use REST tests. For now I think the tests in this PR should suffice. Please let me know if you think otherwise.

**Comment by ywangd** [comment]
The test failures are due to the fact that we don't allow creating (or updating) API keys with remote indices privileges 
https://github.com/elastic/elasticsearch/blob/3f47fe9422aa5166b6b4a793ede5e51c4b794805/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java#L315

This is triggered since the `superuser` role is updated to have remote indices privileges.

I can either wait for #93414 to complete or remove the exception throwing code since it will be removed soon. Please let me know. Thanks!

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java (hunk 1 above)
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:99]: I think mirroring the index privileges (i.e., including `all`) is clearer, and as you point out there are no security concerns here per se.

One thing this raises for me is whether we should (in the future) consider downgrading received privileges on the fulfilling cluster (e.g., `all` maps to `read`, `read_cross_cluster`) to more explicitly enforce the expectation that remote users only ever have read-access. Admins might get lazy and use a remote access API key that has `all`. 

This _should_ be unnecessary and may be overly complex, but it also seems like a good idea to explicitly enforce security assumptions. Certainly not for this PR, and not for now, just jotting this down as a thought.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java [hunk 2]
```diff
@@ -2963,6 +2970,74 @@ public void testPredefinedEditorRole() {
...
+            assertThat(
+                Arrays.stream(superuserRoleDescriptor.getRemoteIndicesPrivileges())
+                    .map(RoleDescriptor.RemoteIndicesPrivileges::indicesPrivileges)
+                    .toArray(RoleDescriptor.IndicesPrivileges[]::new),
+                equalTo(superuserRoleDescriptor.getIndicesPrivileges())
+            );
+            final List<RemoteIndicesPermission.RemoteIndicesGroup> remoteIndicesGroups = superuserRole.remoteIndices()
+                .remoteIndicesGroups();
+            assertThat(remoteIndicesGroups.size(), equalTo(1));
+            assertThat(remoteIndicesGroups.get(0).remoteClusterAliasMatcher().isTotal(), is(true));
```
Comment (n1v0lg) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStoreTests.java:2992]: Neat! TIL about `isTotal`

**Review by n1v0lg — APPROVED** [review]
LGTM, all suggestions & comments are non-blocking.

There is one BWC quirk (I think) that will interact with API keys: if we are in a mixed-node cluster, we should prevent API keys from getting created with remote index privileges, if there are old nodes in the cluster. 

This means that a `superuser` may be unable to create (or update) API keys in a mixed node cluster (since that role now has remote indices). I don't think it makes sense to worry about this in the context of your PR, but rather something for the PR that adds support for API keys.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java (hunk 1 above)
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:99]: > I think mirroring the index privileges (i.e., including all) is clearer, and as you point out there are no security concerns here per se.

👍 

> whether we should (in the future) consider downgrading received privileges on the fulfilling cluster (e.g., all maps to read, read_cross_cluster) to more explicitly enforce the expectation that remote users only ever have read-access. Admins might get lazy and use a remote access API key that has all.

I don't think we should do that. It feels a bit overly smart. The remote access port only allows limited actions. So the privileges themselves are not useful without any actual actions. It is also hard to be complete because privileges can be assigned with actions,e.g. `all` is also `"indices:*", "internal:transport/proxy/indices:*"` and `indices:*` can also be replaced by `indices:admin*` and `indices:data*` and many other variations.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java [hunk 3]
```diff
@@ -81,10 +82,36 @@ public class ReservedRolesStore implements BiConsumer<Set<String>, ActionListene
...
+            : null
     );
     private static final Map<String, RoleDescriptor> RESERVED_ROLES = initializeReservedRoles();
 
+    private static RoleDescriptor.RemoteIndicesPrivileges getRemoteIndicesReadPrivileges(String indexPattern) {
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:104]: I considered that, but overall I prefer a more "matching" local and remote indices privileges. These role descriptors are retrievable via APIs like GetRole. I think it works better if the two sections are "matching". Today if CCS is performed with API keys, it will send entire role descriptor across the network. It is not ideal. But in the same time, I haven't seen it being a concern. So my preference is to keep it as is. Alternatively, I can merge privileges for these indices for both local and remote privilege sections. That might be the best of both worlds?

**Comment by ywangd** [comment]
Thanks for raising the issue about superuser creating API keys! 👍  It's an important issue. I agree that it should be tackled on the API key side. I responded over there (https://github.com/elastic/elasticsearch/pull/93414#discussion_r1100946279).

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (n1v0lg) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/ReservedRolesStore.java:104]: Good call that we want to keep these consistent. Let's just leave as is. 

Merging the indices section for local privileges would change the response structure for the GetRole API; it isn't a breaking change per se, but it feels unnecessarily big, given that we would do it to work around an internal implementation quirk.

**Comment by n1v0lg** [comment]
@ywangd sorry I missed your comment [here](https://github.com/elastic/elasticsearch/pull/93540#issuecomment-1420293638) around the test failure:

I think it makes sense to wait, since removing the exception would also require updating tests so it's be a bigger change but I don't feel strongly. cc @slobodanadamovic

**Comment by ywangd** [comment]
@elasticmachine update branch
